### PR TITLE
Creating Util classes to build a CubeJS Query and sent a Request to a…

### PR DESCRIPTION
**Pull request by freddyucv**
### Proposed Changes
* Create a class to send a Query to a Cube JS Server

https://github.com/dotCMS/core/pull/23895/files#diff-0d9a205dc8b38a352284a9148bef8347fcfdd2a728c68b84d758bc03036600c6R25

* Create a util class to build a CubeJS Query

https://github.com/dotCMS/core/pull/23895/files#diff-c50354f1a06c2e8525d71ddbb42ebb62b376ba7a22d238ff714e76584d3c3674R58

- Create a class to iterate through the CubeJS query data result

https://github.com/dotCMS/core/pull/23895/files#diff-1dadb3f843cc4530acae23891479609d5780ba316cc36d0558c09795d7fecb83R16 

- Create a Mock Http Server Util class

https://github.com/dotCMS/core/pull/23895/files#diff-bb8e6d4042484875cf32b80f746a82828bf3c93ebaa404c732329defc4f9b003R21

- Allow use the CircuitBreakerUrl class in a Test ignoring the IpPrivateSubnet

https://github.com/dotCMS/core/pull/23895/files#diff-8e608dba77e519897741a4a2784b927053dbe8f891a9b9fe485c8eebf5a52474R13

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represent a Result from RUN a CubeJS Query in a CubeJS Server.
+ */
+public class CubeJSResultSet implements Iterable<ResultSetItem> {
+    private List<ResultSetItem> data;
+    private int size;
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java:18]: Given that this is stored by actual `data` object. Do we really need `size` attribute?

File: dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import org.apache.commons.io.IOUtils;
+
+public class MockHttpServerContext {
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java:21]: Thjis class should live outside production code

Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java:21]: I think this is the only way to use it into Integration Test and Unit Test

File: dotCMS/src/main/java/com/dotcms/cube/CubeClient.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * CubeJS Client.
+ */
+public class CubeClient {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/cube/CubeClient.java:25]: Better javadoc

File: dotCMS/src/main/java/com/dotcms/cube/CubeClient.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    private String url;
+
+    public CubeClient(final String url) {
+        this.url = url;
+    }
+
+    public CubeJSResultSet send(final CubeJSQuery query) {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/cube/CubeClient.java:32]: javadoc

File: dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represent a Result from RUN a CubeJS Query in a CubeJS Server.
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java:14]: from running*

File: dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServer.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Mock fro a HttpServer
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServer.java:19]: typo

File: dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java (hunk 2 above)
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java:21]: javadoc

File: dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+import java.util.List;
+import java.util.Map;
+import org.junit.Test;
+
+public class CubeClientTest {
+
+    /**
+     * Method to test: {@link CubeClient#send(CubeJSQuery)}
+     * WHen: Sen a request to Cube JS
```
Comment (dsilvam) [inline dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java:30]: typo

File: dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+     * Should: Return teh right data end send the right Query
```
Comment (dsilvam) [inline dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java:31]: typo

File: dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        } finally {
+            IPUtils.disabledIpPrivateSubnet(false);
+            mockhttpServer.stop();
+        }
+    }
+
+    /**
+     * Method to test: {@link CubeClient#send(CubeJSQuery)}
+     * WHen: Sen a request to Cube JS but the CubeJS Server is down
```
Comment (dsilvam) [inline dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java:104]: typo

File: dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-1dadb3f843cc4530acae23891479609d5780ba316cc36d0558c09795d7fecb83R23

File: dotCMS/src/main/java/com/dotcms/cube/CubeClient.java (hunk 3 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/cube/CubeClient.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-21e76b1df7d7df5fc76a2bab441929ea7c7808dd665f8265eea49548a887b240R22

File: dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/cube/CubeJSResultSet.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-21e76b1df7d7df5fc76a2bab441929ea7c7808dd665f8265eea49548a887b240R22

File: dotCMS/src/main/java/com/dotcms/cube/CubeClient.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/cube/CubeClient.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821

File: dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServer.java (hunk 6 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServer.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-bb8e6d4042484875cf32b80f746a82828bf3c93ebaa404c732329defc4f9b003R19

File: dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/http/server/mock/MockHttpServerContext.java:70]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-de5f0272efb882c609696065290263f8dbfa7d10128686f601eaab3168077b31R21

File: dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java (hunk 7 above)
Comment (freddyucv) [reply dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-07bf99cd1d156dc4ea1bab27b0612fa51e3fa3cb2b52c5c4153366bfb416cc51R30

File: dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java (hunk 8 above)
Comment (freddyucv) [reply dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-07bf99cd1d156dc4ea1bab27b0612fa51e3fa3cb2b52c5c4153366bfb416cc51R31

File: dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java (hunk 9 above)
Comment (freddyucv) [reply dotCMS/src/test/java/com/dotcms/cube/CubeClientTest.java]: done https://github.com/dotCMS/core/pull/23895/commits/951aeb5323d846f15d0c91e766c156be90af7821#diff-07bf99cd1d156dc4ea1bab27b0612fa51e3fa3cb2b52c5c4153366bfb416cc51R104

**Review by victoralfaro-dotcms — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**